asteroid/masknn/convolutional.py

```diff
@@ -211,6 +211,448 @@ def get_config(self):
+        mask_conv = nn.Conv1d(mask_conv_inp, n_src * out_chan, 1)
+        self.mask_net = nn.Sequential(nn.PReLU(), mask_conv)
+        # Get activation function.
+        mask_nl_class = activations.get(mask_act)
+        # For softmax, feed the source dimension.
+        if has_arg(mask_nl_class, "dim"):
+            self.output_act = mask_nl_class(dim=1)
+        else:
+            self.output_act = mask_nl_class()
+
+        self.consistency = nn.Linear(bn_chan, n_src)
+
+    def forward(self, mixture_w):
+        """
+
+        Args:
+            mixture_w (:class:`torch.Tensor`): Tensor of shape
+                [batch, n_filters, n_frames]
+
+        Returns:
+            :class:`torch.Tensor`:
+                estimated mask of shape [batch, n_src, n_filters, n_frames]
+        """
+        batch, n_filters, n_frames = mixture_w.size()
+        output = self.bottleneck(mixture_w)
+        output_copy = output
+
+        skip_connection = 0.0
+        for r in range(self.n_repeats):
+            # Long range skip connection TDCNpp
+            if r != 0:
+                # Transform the input to repeat r-1 and add to new repeat inp
+                output = self.dense_skip[r - 1](output_copy) + output
+                # Copy this for later.
+                output_copy = output
+            for x in range(self.n_blocks):
+                # Common to w. skip and w.o skip architectures
+                i = r * self.n_blocks + x
+                tcn_out = self.TCN[i](output)
+                if self.skip_chan:
+                    residual, skip = tcn_out
+                    skip_connection = skip_connection + skip
+                else:
+                    residual = tcn_out
+                # Initialized exp decay scale factor TDCNpp for residual connections
+                scale = self.scaling_param[r, x - 1] if x > 0 else 1.0
+                residual = residual * scale
+                output = output + residual
+        # Use residual output when no skip connection
+        mask_inp = skip_connection if self.skip_chan else output
+        score = self.mask_net(mask_inp)
+        score = score.view(batch, self.n_src, self.out_chan, n_frames)
+        est_mask = self.output_act(score)
+
+        weights = self.consistency(mask_inp.mean(-1))
+        weights = torch.nn.functional.softmax(weights, -1)
+
+        return est_mask, weights
+
+    def get_config(self):
+        config = {
+            "in_chan": self.in_chan,
+            "out_chan": self.out_chan,
+            "bn_chan": self.bn_chan,
+            "hid_chan": self.hid_chan,
+            "skip_chan": self.skip_chan,
+            "conv_kernel_size": self.conv_kernel_size,
+            "n_blocks": self.n_blocks,
+            "n_repeats": self.n_repeats,
+            "n_src": self.n_src,
+            "norm_type": self.norm_type,
+            "mask_act": self.mask_act,
+        }
+        return config
+
+
+class SuDORMRF(nn.Module):
```

Comment:
Can we have the code of TDCNpp only? There are few other things in this commit.

Reply:
SuDORMRF was not merged ?